[PM-17562] Add documentation for the initial Distributed Events POC (#533)
bitwarden/contributing-docs: docs/getting-started/server/events.md

concrete example of how multiple integrations are enabled by moving to distributed events.

```kroki type=mermaid
graph TD
	  subgraph With RabbitMQ
        B1[EventService]
        B2[RabbitMQEventWriteService]
        B3[RabbitMQ exchange]
        B4[RabbitMqEventRepositoryListener]
        B5[RabbitMqEventHttpPostListener]
        B6[Events Database Table]
        B7[HTTP Server]

        B1 -->|IEventWriteService| B2 --> B3
        B3--> B4 --> B6
        B3--> B5
        B5 -->|HTTP POST| B7
    end

    subgraph Without RabbitMQ
        A1[EventService]
        A2[RepositoryEventWriteService]
        A3[Events Database Table]

        A1 -->|IEventWriteService| A2 --> A3

end
```

### Running the RabbitMQ container

1.  Verify that you've set a username and password in the `.env` file (see `.env.example` for an
    example)

2.  Use Docker Compose to run the container with your current settings:

    ```bash
    docker compose --profile rabbitmq up -d
    ```

    - The compose configuration uses the username and password from the `env` file.
    - It is configured to run on localhost with RabbitMQ's standard ports, but this can be

>       customized in the docker configuration.

Reviewer: ```suggestion
      customized in the Docker configuration.
```

3.  To verify this is running, open `http://localhost:15672` in a browser and login with the
    username and password in your `.env` file.

### Configuring the server to use RabbitMQ for events

1.  Add the following to your `secrets.json` file, changing the defaults to match your `.env` file:

    ```json
    "eventLogging": {
      "rabbitMq": {
        "hostName": "localhost",
        "username": "bitwarden",
        "password": "SET_A_PASSWORD_HERE_123",
        "exchangeName": "events-exchange",
        "eventRepositoryQueueName": "events-write-queue",
      }
    }
    ```

2.  (optional) To use the RabbitMqEventHttpPostListener, specify a queue name and destination URL
    that will receive the POST inside the RabbitMq object:

    ```json
        "httpPostQueueName": "events-httpPost-queue",
        "httpPostUrl": "<HTTP POST URL>",
    ```

    - Tip: [RequestBin](http://requestbin.com/) provides an easy to set up server that will receive
      these requests and let you inspect them.

> 3.  Re-run the powershell script to add these secrets to each Bitwarden project:

Reviewer: ```suggestion
3.  Re-run the PowerShell script to add these secrets to each Bitwarden project:
```

   Identity and web vault)
2. Start the Events project using `dotnet run` or your IDE (note: EventsProcessor is not required
   for self-hosted)

## Distributed events (optional)

> There is a new system which will add support for distributing events via an AMQP messaging system.

Reviewer: ```suggestion
Events can be distributed via an AMQP messaging system which
```

> In the future this will enable new integrations by allowing for a means to subscribe to events via

Reviewer: ```suggestion
enables new integrations by allowing for a means to subscribe to events via
```

Reviewer: Overall change here is to be more declarative about "what is" vs. something coming or "new".

messaging stream.

> As an initial proof of concept, there is an optional RabbitMQ implementation that refactors the way

Reviewer: 🎨 Same kind of change. These docs describe what's available.

events are handled when running locally or self-hosted. Instead of writing directly to the `Events`

> table via the `EventsRepository`, it will broadcast each event via a RabbitMQ exchange. A new

Reviewer: 🎨 Change to present tense everywhere too.

`RabbitMqEventRepositoryListener` then subscribes to the RabbitMQ exchange and writes to the
`Events` table via the `EventsRepository`. The end result is the same (events are stored in the
database), but this allows for other integrations to subscribe.

To illustrate this, there is also a `RabbitMqEventHttpPostListener` which subscribes to the RabbitMQ
events exchange and `POST`s each event to a configurable URL. This is meant to be a simple, concrete
example of how multiple integrations are enabled by moving to distributed events.

```kroki type=mermaid
graph TD
	  subgraph With RabbitMQ
        B1[EventService]
        B2[RabbitMQEventWriteService]
        B3[RabbitMQ exchange]
        B4[RabbitMqEventRepositoryListener]
        B5[RabbitMqEventHttpPostListener]
        B6[Events Database Table]
        B7[HTTP Server]

        B1 -->|IEventWriteService| B2 --> B3
        B3--> B4 --> B6
        B3--> B5
        B5 -->|HTTP POST| B7
    end

    subgraph Without RabbitMQ
        A1[EventService]
        A2[RepositoryEventWriteService]
        A3[Events Database Table]

        A1 -->|IEventWriteService| A2 --> A3

end

```

> To enable the new RabbitMQ-based event stream, take the following steps:

Reviewer: ⛏️ I'd delete this whole line as the need to take action is implied.

### Running the RabbitMQ container

1.  Verify that you've set a username and password in the `.env` file (see `.env.example` for an
    example)

2.  Use Docker Compose to run the container with your current settings:

    ```bash
    docker compose --profile rabbitmq up -d
    ```

> 3.  This will run the RabbitMQ container with your username and password on localhost with the

Reviewer: ⛏️ This isn't a step but more detail for the prior step.
